src/etl/impl/LedgerPublisher.h

```diff
@@ -0,0 +1,253 @@
+    mutable std::shared_mutex publishTimeMtx_;
+
+    std::optional<uint32_t> lastPublishedSequence_;
+    mutable std::shared_mutex lastPublishedSeqMtx_;
+
+public:
+    /**
+     * @brief Create an instance of the publisher
+     */
+    LedgerPublisher(
+        boost::asio::io_context& ioc,
+        std::shared_ptr<BackendInterface> backend,
+        std::shared_ptr<SubscriptionManager> subscriptions,
+        SystemState const& state)
+        : publishStrand_{ioc}, backend_{backend}, subscriptions_{subscriptions}, state_{std::cref(state)}
+    {
+    }
+
+    /**
+     * @brief Attempt to read the specified ledger from the database, and then publish that ledger to the ledgers
+     * stream.
+     *
+     * @param ledgerSequence the sequence of the ledger to publish
+     * @param maxAttempts the number of times to attempt to read the ledger from the database. 1 attempt per second
+     * @return whether the ledger was found in the database and published
+     */
+    bool
+    publish(uint32_t ledgerSequence, std::optional<uint32_t> maxAttempts)
+    {
+        log_.info() << "Attempting to publish ledger = " << ledgerSequence;
+        size_t numAttempts = 0;
+        while (not state_.get().isStopping)
+        {
+            auto range = backend_->hardFetchLedgerRangeNoThrow();
+
+            if (!range || range->maxSequence < ledgerSequence)
+            {
+                log_.debug() << "Trying to publish. Could not find "
+                                "ledger with sequence = "
+                             << ledgerSequence;
+
+                // We try maxAttempts times to publish the ledger, waiting one second in between each attempt.
+                if (maxAttempts && numAttempts >= maxAttempts)
+                {
+                    log_.debug() << "Failed to publish ledger after " << numAttempts << " attempts.";
+                    return false;
+                }
+                std::this_thread::sleep_for(std::chrono::seconds(1));
+                ++numAttempts;
+                continue;
+            }
+            else
+            {
+                auto lgr = Backend::synchronousAndRetryOnTimeout(
+                    [&](auto yield) { return backend_->fetchLedgerBySequence(ledgerSequence, yield); });
+
+                assert(lgr);
+                publish(*lgr);
+
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @brief Publish the passed in ledger
+     *
+     * All ledgers are published thru publishStrand_ which ensures that all publishes are performed in a serial fashion.
+     *
+     * @param lgrInfo the ledger to publish
+     */
+    void
+    publish(ripple::LedgerInfo const& lgrInfo)
+    {
+        boost::asio::post(publishStrand_, [this, lgrInfo = lgrInfo]() {
+            log_.info() << "Publishing ledger " << std::to_string(lgrInfo.seq);
+
+            if (!state_.get().isWriting)
+            {
```

Comment:
Can we move the cache update to ETLService, i can see two points:
1 make the functionality more separate.Publisher just publish to subscribers. We also can remove shared state isWriting.
2 I remember in future we want to  remove backend cache() function.

Reply:
Would love to but that would mean it's no longer done from the `publishStrand_` and who knows what kind of side effect that will have. I'd rather do that in the future just so we don't change too much flow/logic with this pr.

Reply:
Ha i see